### DISCUSSION TITLE
Effect over irregularly distributed LEDs in plane

### DISCUSSION THREAD

**Discussion by zdar** [PlummersSoftwareLLC/NightDriverStrip#15] [Q&A]
Hi,
I have led "strip" which has LEDs randomly distributed in space. I know spatial coordinates (XY) for each of LED in strip.
Is there a way to use NightDriverStrip effects to work with such strip? 
If I understand code well they work only over regular LED matrix :-(

**Comment by ohnoitsalobo** [comment]
Is this [XY map generator](https://macetech.github.io/FastLED-XY-Map-Generator/) close to what you're looking for?

Basically you'd write your code for a regular matrix, but call this mapping function first before actually sending it out to the LEDs. Since it's nothing but a lookup table, it's extremely fast.

**Reply by zdar** [reply]
Unfortunatelly this it does not help that much. But I may use the idea from it as temporary workabout. Thanks for the link!

There are two problems with this approach:
- The coordinates of my pixels are highly irregular. That means the fake matrix would need to be quite large and very sparse. If I wan't to prevent distance distortion it would mean even larger fake matrix.
- I do not have "strip" organised. That means the numbering is not going continous from top left as tool anticipates. However this could be easilly solved.

What I'd like to do is to be able to ask effect for color at particular XY coordinates (those are not matrix coordinates but spatial coordinates) and set color of pixel by that. Nowdays it is reverse - the effect chooses pixels to set.

**Reply by ohnoitsalobo** [reply]
> The coordinates of my pixels are highly irregular. That means the fake matrix would need to be quite large and very sparse. If I want to prevent distance distortion it would mean even larger fake matrix.

I totally understand, but even an extremely large matrix lookup is insanely fast on processors like the ESP32. It's a bit of work up-front - but unless the pixels are going to move around, it should be pretty much 100% reliable and won't require any changes to the NightDriver code.

You should provide some sort of visual example so that we can see exactly what problem you're trying to solve; typically, the fastest way is to take the pixel data and re-map it like this.

It even works on a curved surface: https://www.youtube.com/watch?v=zYZkbfmCuEY

**Reply by zdar** [reply]
I'm working on PCB based stackable tile set (triangles and hexagons). Nanoleaf/moodlite alternative using cheap PCB+SMD assembly in China.

![tile](https://user-images.githubusercontent.com/1691703/137202516-8d2a33cf-0d8a-447f-afac-430e26b71bc8.jpg)

Each tile has set of LEDs (triangle 3, hexagon 18) and tiles can be to connected to various patterns. This can in some instances lead to quite sparse matrix. But as You said there should be enough memory for that so I'll give it a try :-)

**Reply by Joshfindit** [reply]
Very cool. Are the PCBs flat on the bottom/back? Or do you have “stick” PCBs on the inside of the perimeter?

**Reply by zdar** [reply]
Triangle has a flat PCB. Hexagon has a perimeter "string" of PCBs. Both can fit into 10x10cm project for manufacturing (hexagon PCB needs to be cut into pieces and soldered into ring).

![triangle pcb](https://user-images.githubusercontent.com/1691703/137217080-59cc9534-55b6-44f9-8c91-b3d6ac3641db.png)
![hexagon pcb](https://user-images.githubusercontent.com/1691703/137217090-5cd20a3b-e1dd-4a13-9613-d5f34b0e019f.png)

**Comment by davepl** [comment]
On an ESP32 it can update a chain of 1140 LEDs about 24 times a second.  Faster if fewer LEDs, and it can do that on 8 channels at once!

Do the panels feature concentric rings?  How do you envision drawing into them?  Like what is your dream API to be able to set a pixel on one of them?

- Dave

**Reply by zdar** [reply]
Panels have mostly "rings". But topology can be more complicated as I can fit six triangles inside hexagon ring. At that moment there is set of adjacent rings (non concentric) inside big one. Anway I'll try the fake matrix approach once I have more panels ready. The LED count (including fake ones) should not be high for ESP32.

API - it is more about kind of reverse approach. At this moment everyhing is setup in a way that effect draws on a matrix canvas. I'd like to do it in opposite way. An iterator over LED list (matrix) will call efffect API to ask about color at specific spatial coordinate for each LED. This way even 3D should not be a problem. I can even anticipate a use a coordinate filter chain (symetry effect, radial effects, 3D->2D->1D mapping, ...) in this in future.

**Comment by robertlipe** [comment]
This conversation kind of coasted to a halt without ever being closed.

THe mapping solution suggested by @ohnoitsalobo is certainly popular in WLED. There's not an implementation in that in NightDriver. I've flirted with needing it a few times and certainly enables some interesting approaches.

The specific hardware that zdar is asking about, if I understand it correctly, is wacky enough that evne the common generators like:
https://jasoncoon.github.io/led-mapper/ (rocks!)
https://macetech.github.io/FastLED-XY-Map-Generator/
https://intrinsically-sublime.github.io/FastLED-XY-Map-Generator/
https://intrinsically-sublime.github.io/WLED-Arbitrary-Path-Ledmap.json-Generator/
https://private-user-images.githubusercontent.com/1432368/325113956-47e4c878-596c-4526-a427-d6f6f9a4f9e2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYzODM4OTIsIm5iZiI6MTcyNjM4MzU5MiwicGF0aCI6Ii8xNDMyMzY4LzMyNTExMzk1Ni00N2U0Yzg3OC01OTZjLTQ1MjYtYTQyNy1kNmY2ZjlhNGY5ZTIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MTVUMDY1OTUyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTZjZjg1YzkxYWYwMGU1ZjMzY2ViMzY5MzIxODEwYWY2Yzc2NzM4NzQ3OGFhNGU0ZDQwYzVkNWQ0M2I2NDRkYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.4ndun49B_uOhm44mup8hJpWSlHOtq22V8Nz9lbSmN24
- and a TON of others have implemented - would still be a real pain to lay out a wall full of panels as shown in https://github.com/PlummersSoftwareLLC/NightDriverStrip/discussions/15#discussioncomment-1473072 would be a real pain to implement in ANY software, particularly one that doesn't have great level of indirection beyond the pixel buffers into 2/3D output space anyway.

It's easy to imagine a magnetic/velcro approach that lets someone stick zany combinations of these things onto a wall, but it'd be a real pain to represent these in any kind of memory-efficient way that (A) would be a reasonable thing to implement in a UI at all and (B) would allow things like X/Y mappings to that an "explosion" in the bottom center arced to outer branches and then gravity-fell to the bottom or sparked off. The brute force would be to have a wall-sized X/Y pixel mapping that's probably sparse (the user can't afford a wall of HUB75s - at some point, you just use monitors. :-) ) and just has a multi-level map that subsegments regional pixels into an X/Y matrix to allow normal patterns to work on them.

For physically large boundaries (a wall's worth of pixels) this becomes a nightmare's worth of data to manage, even cleverly, in a humble 320k ESP32 and certainly not a walk in the park for even an 8MB model that's currently the top of the value line of ESP32. (Three years later, there's talk of an ESP32 that'll do 64MB, but none have been built and as such, it's not something now supported.) If you're willing to throw more MCUs at the problem and have each "neighborhood" of pixels controlled by a local HOA and have them all chattering to each other over something like ESP-NOW or even (gasp) wires, frame rates get  easier, but defining an overall representation of the whole thing doesn't. A 3,000x3,000px wall is hard, even if lots of the pixels are missing.



It's probably worth reframing the question, at least as I understand it, as to what kind of aesthetically pleasing effects could be rendered onto such a configuration. You might not be able to do a "stars and stripes waving in the wind" or "gif renderer" in a way that's practical while allowing arbitrary geometry there, but it might be fun for someone with art skills to explore what's attractive with panels like these in a way that makes sense given the realities of computer science/hardware expense that we're all currently bound.  Then again, maybe it's the kind of thing that gets rendered on a Real Computer™ and shipped over WiFi to the blinky farm on the wall. That's something NightDriverLED and something like XLights, ledfx.app,  or another renderer could pull off.

I could double-check the math about 1140 LEDs @ 24 fps, but just from memory, that's about the commonly accepted numbers for the reality of the bandwidth of WS281x protocol. It's an 800khz NRZ protocol with about three subframes per bit, 24 bits of color that's 2 colors * 256 brightness levels, etc. That's just about the point where the "one wire" (I hate that term) approach runs out of gas as bound by those paramaters. Going faster involves changing LED strands, signalling protocols, etc. It moves from dollar-store hardware to complicated things - lots of problems can be solved by applying money if that's the goal. :-)


Do these panels actually exist now? What software supports them Really Well? What could NightDriverLED do better?

I'll have to say that if you're willing to roll something in code, NightDriverLED has really good underpinnings to let you engineer/build code and products like this.

If you're looking to paw at something without actually writing code, WLED probably gets you to "pretty patterns on a wall" faster (maybe each "strip" is an ordered "segment" that you manually place, etc.) but you probably run into hard limits that are harder to remove later.
